r/src/compute.cpp

```diff
@@ -574,3 +576,186 @@ SEXP compute__CallFunction(std::string func_name, cpp11::list args, cpp11::list
+            if (!result->type()->Equals(scalar->type)) {
+              arrow::Datum out =
+                  ValueOrStop(arrow::compute::Cast(scalar, result->type()->Copy()));
+              std::shared_ptr<arrow::Scalar> out_scalar = out.scalar();
+              scalar.swap(out_scalar);
+            }
+
+            // make sure we assign the type that the result is expecting
+            if (result->is_scalar()) {
+              result->value = std::move(scalar);
+            } else {
+              auto array = ValueOrStop(arrow::MakeArrayFromScalar(
+                  *scalar, span.length, context->memory_pool()));
+              result->value = std::move(array->data());
+            }
+          } else {
+            cpp11::stop("arrow_scalar_function must return an Array or Scalar");
+          }
+        },
+        "execute scalar user-defined function");
+  }
+};
+
+// [[arrow::export]]
+void RegisterScalarUDF(std::string name, cpp11::sexp func_sexp) {
+  cpp11::list in_type_r(func_sexp.attr("in_type"));
+  cpp11::list out_type_r(func_sexp.attr("out_type"));
+  R_xlen_t n_kernels = in_type_r.size();
+
+  if (n_kernels == 0) {
+    cpp11::stop("Can't register user-defined function with zero kernels");
+  }
+
+  // compute the Arity from the list of input kernels
+  std::vector<int64_t> n_args(n_kernels);
+  for (R_xlen_t i = 0; i < n_kernels; i++) {
+    auto in_types = cpp11::as_cpp<std::shared_ptr<arrow::Schema>>(in_type_r[i]);
+    n_args[i] = in_types->num_fields();
+  }
+
+  int64_t min_args = *std::min_element(n_args.begin(), n_args.end());
+  int64_t max_args = *std::max_element(n_args.begin(), n_args.end());
+
+  // We can't currently handle variable numbers of arguments in a user-defined
+  // function and we don't have a mechanism for the user to specify a variable
+  // number of arguments at the end of a signature.
+  if (min_args != max_args) {
+    cpp11::stop(
+        "User-defined function with a variable number of arguments is not supported");
+  }
+
+  arrow::compute::Arity arity(min_args, false);
+
+  // The function documentation isn't currently accessible from R but is required
+  // for the C++ function constructor.
+  std::vector<std::string> dummy_argument_names(min_args);
+  for (int64_t i = 0; i < min_args; i++) {
+    dummy_argument_names[i] = "arg";
+  }
+  const arrow::compute::FunctionDoc dummy_function_doc{
+      "A user-defined R function", "returns something", std::move(dummy_argument_names)};
+
+  auto func =
+      std::make_shared<arrow::compute::ScalarFunction>(name, arity, dummy_function_doc);
+
+  for (R_xlen_t i = 0; i < n_kernels; i++) {
+    auto in_types = cpp11::as_cpp<std::shared_ptr<arrow::Schema>>(in_type_r[i]);
+    cpp11::sexp out_type_func = out_type_r[i];
+
+    std::vector<arrow::compute::InputType> compute_in_types;
+    for (int64_t j = 0; j < in_types->num_fields(); j++) {
+      compute_in_types.push_back(arrow::compute::InputType(in_types->field(i)->type()));
+    }
```

Comment:
May be?

```suggestion
    std::vector<arrow::compute::InputType> compute_in_types(in_types->num_fields());
    for (int64_t j = 0; j < in_types->num_fields(); j++) {
      compute_in_types.emplace_back(arrow::compute::InputType(in_types->field(j)->type()));
    }
```
`j` is never used, is it a mistake using `i` here?

Reply:
👍